### DISCUSSION TITLE
Allow set to override and implement init

### DISCUSSION THREAD

**Discussion by YairHalberstadt** [dotnet/csharplang#4031] [Language Ideas]
The following scenarios should be legal:

```csharp
public interface I
{
    public int P { get; init; }
}

public class C : I
{
    public int P { get; set; }
}

public class Base
{
    public abstract int P { get; init; }
}

public class Derived : Base
{
    public override int P { get; set; }
}
```

**Comment by liamlim** [comment]
I like this proposal and it solves all problems of #4030 that I have. Do I understand correctly that following code would be legal?

```csharp
public class Base
{
    public abstract int P { get; init; }
}

public class Derived : Base
{
    public void SetP() => P = 42;

    public override int P { get; private set; }
}
```

```csharp
var d = new Derived 
{
  P = 42
};
d.SetP();
```

**Reply by YairHalberstadt** [reply]
I don't think so - we would need a new feature where you could have something both `init`, and `private set`.

The problem is `init` is both wider than `private set` because it allows setting the property publicly in object initializer, and less wide, since it doesn't allows setting it at all outside object initializers.

**Reply by liamlim** [reply]
I imagined init property as setter with init accessibility modifier which is less accessible than private because it allows to assign only during creation of object. What is wrong about this implementation? I agree about less wide part but I don't agree about wider part. We cannot call object initializer more times to change assigned value thus I don't agree that the value is publicly available

**Reply by liamlim** [reply]
Compare following (I marked differences of both definitions with bold)

```csharp
class A
{
  public A(int a)
  {
    A = a;
  }

  public A { get; private set; }
}
```
With this definition, we can assign to **private** setter using public **constructor** during construction of an object. This DOES NOT mean that the setter is publicly available because **constructor** can be called only once.

```csharp
class A
{
  public A { get; init /* set */; }
}
```
With this definition, we can assign to **init** setter using public **property** during construction of an object. This DOES NOT mean that the setter is publicly available because **initializer** can be called only once.

**Comment by JeroenBos** [comment]
>     public class Base
>     {
>         public abstract int P { get; init; }
>     }

Currently, class `Base` can assume that property `P` won't change after the initializer, which in a useful invariant.

This proposal would break that invariant.

**Reply by YairHalberstadt** [reply]
> Currently, class Base can assume that property P won't change after the initializer, which in a useful invariant.
>
> This proposal would break that invariant.

No it can't:

```csharp
public class Base
{
    public abstract int P { get; init; }
}

public class Derived : Base
{
    public PBackingProperty { get; set; }
    public override int P { get => PBackingProperty; init => PBackingProperty = value; }
}
```
